Allow runtimeconfig StartupHooks and Environment StartupHooks to both be present

tomdegoede opened the pull request:
When building/running an app via [our Bazel build implementation](https://github.com/AFASResearch/rules_dotnet) we rely on a startup hook for assembly loading. We specify this in the runtimeconfig.json as a configProperty STARTUP_HOOKS.

When another startup hook is bootstrapped via the environment variable DOTNET_STARTUP_HOOKS this results in the following error:
```
Duplicate runtime property found: STARTUP_HOOKS
It is invalid to specify values for properties populated by the hosting layer in the the application's .runtimeconfig.json
```

In our case this environment variable is added by the IDE to handle net6.0 HotReloading.

Technically it is allowed to specify multiple startup hooks with a delimiter. This pull request resolves above error by merging startup hooks when they are specified in both locations.

Since I rarely visit C++ a detailed review is very welcome. I could not find any existing tests for this code path to extend/replicate but have tested that it resolves my issue. I have also taken the opportunity to fix some log_duplicate_property_errors that mistakenly list the StartUpHooks property key.

tomdegoede committed:
Allow configProperties "STARTUP_HOOKS" and env "DOTNET_STARTUP_HOOKS" to both be present

tomdegoede committed:
Fix duplicate_property_error property names to reflect the actual duplicate

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=61461) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=61461) before we can accept your contribution.<br/><br/>:x: tomdegoede [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=61461)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=61461) it.</sub>

vitek-karas commented on src/native/corehost/hostpolicy/hostpolicy_context.cpp:
```diff
@@ -281,11 +281,20 @@ int hostpolicy_context_t::initialize(hostpolicy_init_t &hostpolicy_init, const a
         {
-            log_duplicate_property_error(coreclr_property_bag_t::common_property_to_string(common_property::StartUpHooks));
-            return StatusCode::LibHostDuplicateProperty;
+#ifdef TARGET_UNIX
+            char separator = ':';
+#else
+            char separator = ';';
+#endif // TARGET_UNIX
```
`PATH_SEPARATOR` (defined in `pal.h`)

vitek-karas commented on src/native/corehost/hostpolicy/hostpolicy_context.cpp:
```diff
@@ -297,7 +306,7 @@ int hostpolicy_context_t::initialize(hostpolicy_init_t &hostpolicy_init, const a
         if (!coreclr_properties.add(common_property::BundleProbe, ptr_stream.str().c_str()))
         {
-            log_duplicate_property_error(coreclr_property_bag_t::common_property_to_string(common_property::StartUpHooks));
+            log_duplicate_property_error(coreclr_property_bag_t::common_property_to_string(common_property::BundleProbe));
```
❤️ - thank you!

vitek-karas reviewed (commented):
Looks good on the product side. We should add a test. The tests for startup hooks are here: https://github.com/dotnet/runtime/blob/main/src/installer/tests/HostActivation.Tests/StartupHooks.cs

(The host tests are a bit tricky to run - please take a look here: https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/host/testing.md)

Thanks a lot for this!

vitek-karas commented:
Question unrelated to the PR:
What is the scenario where you want to bake the startup hook into the `.runtimeconfig.json`?

tomdegoede commented:
> Question unrelated to the PR: What is the scenario where you want to bake the startup hook into the `.runtimeconfig.json`?

Our build tool (Bazel) doesn't copy dependencies to the output directory but instead lists their paths in a App.runfiles_manifest. To properly resolve the dll's at runtime we add a startup hook that registers Resolve handlers on AssemblyLoadContext.Default. In many scenarios we could register this via the environment variable, but in some cases (such as running tests from the IDE) this would be quite a hassle. The App.runtimeconfig.json of the target binary is always considered and thus fixes assembly loading in all cases. It also allows us to run `bin/App.exe` without requiring any additional environment variables, thus not requiring a wrapper or .bat file.

I'll look into the feedback. Thanks for the quick reply!

tomdegoede committed:
fixup! Allow configProperties "STARTUP_HOOKS" and env "DOTNET_STARTUP_HOOKS" to both be present

tomdegoede commented:
> Looks good on the product side. We should add a test. The tests for startup hooks are here: https://github.com/dotnet/runtime/blob/main/src/installer/tests/HostActivation.Tests/StartupHooks.cs
> 
> (The host tests are a bit tricky to run - please take a look here: https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/host/testing.md)
> 
> Thanks a lot for this!

@vitek-karas I have added tests and reused the predefined PATH_SEPARATOR.

vitek-karas reviewed (approved):
Looks great - thanks a lot!

tomdegoede commented:
> Hmm - thinking about the order. I wonder if it would be better to run the env. variable hooks first.
> 
> In other places (for example [roll forward](https://github.com/dotnet/runtime/blob/main/docs/design/features/framework-version-resolution.md#roll-forward)) the env. variable takes precedence over runtime config.
> 
> Typically the order of startup hooks should not matter as they should not make assumptions about other possible startup hooks. But in reality that's not always the case, as typically the thing starting the process also "owns" the startup hook and sets up the env. variable accordingly.
> 
> I lean toward having the startup hooks from the env. variable run BEFORE those from the runtime config.
> 
> @sbomer, what do you think?

I think the env hook is already executed first since I append the existing config_startup_hooks to the env string_t. This was however not an explicit decision, nor do the tests currently validate it.

sbomer commented:
Agreed, the env var should win, since the original design was addressing situations in which modifying deployed bits in the runtimeconfig was not an option.

vitek-karas commented:
So we're in agreement, env. variable startup hooks should run before the runtime config ones.
@sbomer I just want to make sure we are in agreement:
* Runtime config registers a startup hook "RuntimeConfigHook"
* Env. variable is set to a startup hook "VariableHook"

Running the app will run both startup hooks, first the "VariableHook" and then the "RuntimeConfigHook" (and then the `Main`).

The code in this PR already works this way. Testing it will be a bit tricky, we don't have test helpers to test for order of messages in the output. The simplest way would be to grab the entire test output and implement the validation manually (search for first, and then search again from that point for the second).

I think we should try to add a test for this, it's an important detail and it's easy to break in the future if the code changes. Please add a comment into the code that the order is intentional.

tomdegoede committed:
Enforce env startup hook precedence over runtimeconfig startup hooks

tomdegoede commented:
@vitek-karas I have updated the testcase to consider stdOut order and added a comment in the code path. Please let me know if there is anything else to improve! Thanks :)

vitek-karas commented on src/installer/tests/HostActivation.Tests/StartupHooks.cs:
```diff
@@ -147,16 +147,20 @@ public void Muxer_activation_of_RuntimeConfig_And_Environment_StartupHooks_Succe
-                .And.HaveStdOutContaining("Hello from startup hook!")
-                .And.HaveStdOutContaining("Hello from startup hook with dependency!")
-                .And.HaveStdOutContaining("Hello World");
+                .And.HaveStdOutMatching("Hello from startup hook with dependency!" +
+                                        whitespacePattern +
+                                        "Hello from startup hook!" +
+                                        whitespacePattern +
+                                        "Hello World");
```
This is unfortunately not 100% reliable.
We've seen tests fail because there are multiple threads running at the same time (bread-crumb writer) which can also print to output.
See https://github.com/dotnet/runtime/issues/57243. I don't know 100% if that can happen here as well... but I'd rather make the test resilient to that.

We could probably change the regex to accept any text in between to "fix" this.

vitek-karas replied:
I have to apologize - I should have mentioned this problem beforehand... sorry.

tomdegoede replied:
ah np I understand. Thanks for letting me know.

tomdegoede replied:
I have slightly changed the pattern and added a comment.

vitek-karas commented on src/native/corehost/hostpolicy/hostpolicy_context.cpp:
```diff
@@ -284,6 +284,8 @@ int hostpolicy_context_t::initialize(hostpolicy_init_t &hostpolicy_init, const a
         if (coreclr_properties.try_get(common_property::StartUpHooks, &config_startup_hooks))
         {
+            // env startup hooks shoold have precedence over config startup hooks
+            // therefore append config_startup_hooks ANFTER startup_hooks
```
```suggestion
            // therefore append config_startup_hooks AFTER startup_hooks
```

vitek-karas commented on src/installer/tests/HostActivation.Tests/StartupHooks.cs:
```diff
@@ -105,6 +106,64 @@ public void Muxer_activation_of_Multiple_StartupHooks_Succeeds()
+            var startupHook2Dll = startupHook2Fixture.TestProject.AppDll;
+
+            // include any char to counter output from other threads such as in #57243
+            const string whitespacePattern = @"[\r\n\s.]*";
```
Nit: should not be called "whitespace" anymore I guess.

tomdegoede replied:
I changed it to wildcard :)

vitek-karas reviewed (approved):
@sbomer, can you please check the code as well (just to make sure, since this is both fix and tiny design change in one :wink:)

sbomer reviewed (approved):
LGTM, thank you!

tomdegoede commented:
@vitek-karas small beginners question: will this end up in an upcoming patch version for net6 or only in the next net7? Ideally our issue is resolved in net6.

vitek-karas commented:
Currently the `main` is for .NET 7 - so that will happen automatically.
.NET 6 would be harder - we would need to come up with a good reason to do so and all of the risk versus reward discussion.

On its own I like this - specifically because of hot-reload using startup hooks. That said, this is the first time I've heard anybody specifying startup hooks in the runtimeconfig. I'll ask others how they would feel about this.. currently for me, it probably doesn't feel important enough to include in servicing. But there might be different expectations for some of the future servicing releases of .NET 6, so maybe that will change.

agocke commented:
This looks like a good fix to me. Low risk, addresses an oversight in startup hooks specified through two different mechanisms.

agocke commented:
/backport to release/6.0

agocke commented:
uhhh ?? @terrajobst assistance with the above?

tomdegoede commented:
@agocke @terrajobst would be great if backporting is accepted! Thanks!

terrajobst commented:
> @agocke is not a repo collaborator, backporting is not allowed.

Your user is an org member and has push permissions. I've pinged @mmitche with details.

terrajobst commented:
@tomdegoede 

> @agocke @terrajobst would be great if backporting is accepted! Thanks!

Yeah, we're just figuring out how to convince our tooling first :-)

agocke commented:
/backport to release/5.0

agocke commented:
/backport to release/6.0

agocke commented:
/backport to release/6.0
